---
title: "How to remove unwanted variation from transcriptomics data using RUV4"
author: Momeneh (Sepideh) Foroutan and Marie Trussart
output:
  github_document:
    toc: yes
  html_notebook:
    number_sections: yes
    toc: yes
    toc_float: yes
    code_folding: "hide"
    fig_caption: yes
---


# Data description
In this tutorial, we aim to obtain differentially expressed genes (DEGs) between two biological conditions in an integrated data. This integrated data set is generated by combining 10 microarray studies on control and TGFb-treated samples (for more information about the data sets, see [A Transcriptional Program for Detecting TGFÎ²-Induced EMT in Cancer](http://mcr.aacrjournals.org/content/15/5/619) by Foroutan M et al, 2017).\ 

First, we assess the presence of unwanted variation in the integrated data which combines differents studies and platforms. Second, we show how to use RUV4 to remove unwanted variation and detect DEGs comparing control to TGFb-treated samples. Finally, we assess whether or not running RUV4 method helps, where we include the analysis of unadjusted data.
```{r load-libs}
library(ruv)            ## for applying RUV method
library(limma)          ## for vennDiagram()
library(ggplot2)        ## for data visualisation
library(knitr)          ## for Kable()
library(shiny)          ## for shiny app
library(colourpicker)
```

The integrated data introduced above have been split into two data sets: dataset A and dataset B, each containing different studies, platforms and tissues. We will explore and normalise these two data sets separately in order to compare the results obtained by the normalisation method RUV4.

Read in the two integrated datasets.
```{r load-data}
datasetA<- read.table("expr_10data_datasetA.txt", header = T, sep = "\t")
datasetB<- read.table("expr_10data_datasetB.txt", header = T, sep = "\t")
```

Look at the each data sets A and B, then make a matrix for each these data sets where the row names are gene IDs.
```{r explore-data}

kable(datasetA[1:4,1:4])
kable(datasetA[1:4,1:4])
mA<-datasetA[,2:dim(datasetA)[2]]
mB<-datasetB[,2:dim(datasetB)[2]]
row.names(mA)<- datasetA[,1]
row.names(mB)<- datasetB[,1]
mA<- as.matrix(mA)
mB<- as.matrix(mB)
```

Look at the information related to each dataset including the name of the studies, types of platform, treatment, and tissue:
```{r load-info-study}
info_datasetA <- read.table("info_10data_datasetA.txt", sep="\t", header=T)
kable(info_datasetA)
info_datasetB <- read.table("info_10data_datasetB.txt", sep="\t", header=T)
kable(info_datasetB)
```

# Assessment of unwanted variation in the data
Here we perform some exploratory analyses on the integrated data sets A and B to assess the presence of unwanted variation in each dataset.

## RLE plot
We start by looking at the RLE plots in dataset A, coloured by study, platform and tissue. In a "good" RLE plot, the median of all boxplots (i.e. samples) are around zero, while high deviations from zero show the presence of unwanted variation. For more information about RLE plot see  paper [RLE Plots: Visualising Unwanted Variation in High Dimensional Data](https://arxiv.org/abs/1704.03590), by Luke C. Gandolfo, Terence P. Speed, 2017.\
In the current example, it's important to note that in some cases, different studies are confounded with different platforms and tissues, and therefore there is no way to identify how much of the unwanted variation come from each of these factors. Such situations must be avoided when designing an experiment, however, it often happens when we perform data integration. For the purpose of this tutorial, we only consider "study" as the source of unwanted variation.\

```{r rleplots-datasetA}
## Transpose the expression matrix so that we have genes in columns and dataset in rows
YA <- t(mA)
## Plot RLE coloured by study
ruv_rle(YA, info_datasetA$study, ylim = c(-5,5))
## Plot RLE coloured by platform
ruv_rle(YA, info_datasetA$platform, ylim = c(-5,5))
## Plot RLE coloured by tissue
ruv_rle(YA, info_datasetA$tissue, ylim = c(-5,5))
```

Similarly, we look at the RLE plots in dataset B coloured by study, platform and tissue:
```{r rleplots-datasetB}
YB <- t(mB)
## Plot RLE coloured by study
ruv_rle(YB, info_datasetB$study, ylim = c(-5,5))
## Plot RLE coloured by platform
ruv_rle(YB, info_datasetB$platform, ylim = c(-5,5))
## Plot RLE coloured by tissue
ruv_rle(YB, info_datasetB$tissue, ylim = c(-5,5))
```

## PCA plot
In transcriptomics applications, one of the most utilized exploratory plots is the multi-dimensional scaling (MDS) plot or a principal component analysis (PCA) plot. To assess the presence of unwanted variation in each dataset, we use PCA plots to show similarities between samples measured in an unsupervised way. Ideally, samples should be clustered together according to the treatment (i.e. the biological factor of interest). Here, we see that samples are rather clustered by studies (i.e. unwanted variation) in both data sets A and B.\

```{r mds-plot-datasetA}
gg_additions <- list(aes(color = info_datasetA$study, 
                         shape = info_datasetA$treatment, 
                         size = 5, alpha = .7),
                     labs(color = "Study", shape = "Treatment"),
                     scale_size_identity(guide = "none"),
                     scale_alpha(guide = "none"),
                     theme(legend.text = element_text(size = 12),
                           legend.title = element_text(size = 16)),
                     guides(color = guide_legend(override.aes = list(size = 4)),
                            shape = guide_legend(override.aes = list(size = 4))))
options(repr.plot.width = 8, repr.plot.height = 6)
ruv_svdplot(YA) + gg_additions 
```

```{r mds-plot-datasetB}
gg_additions <- list(aes(color = info_datasetB$study,
                         shape = info_datasetB$treatment,
                         size = 5, alpha = .7),
                     labs(color = "Study",shape = "Treatment"),
                     scale_size_identity(guide = "none"),
                     scale_alpha(guide = "none"),
                     theme(legend.text = element_text(size = 12),
                           legend.title = element_text(size = 16)),
                     guides(color = guide_legend(override.aes = list(size = 4)),
                            shape = guide_legend(override.aes = list(size = 4))))
options(repr.plot.width=8, repr.plot.height=6)
ruv_svdplot(YB) + gg_additions #
```


# Remove batch effects using RUV-4
There are several RUV methods for removing unwanted variation in order to obtain DEGs, including RUV-2, RUV-4, RUV-inv and RUV-rinv. For RUV-4 we can select different values for **dimension of unwanted variation (k)**, while for RUV-inv and RUV-rinv we don't need to change k as it is set to be the maximum value. RUV-inv is recommended when we have large number of control genes (~1000), while RUV-rinv is more appropriate with small number of control genes (~60).\

Negative control genes are genes which are not associated with the biological factor of interest. Selection of negative control genes is very important. Examples of negative control genes are the *spike-in controls* or the *housekeeping (HK) genes*. It is also possible to define *empirical negative control genes* using an iterative approach, however, it is only recomended if (i) the initial negative control genes are not very good or there are only a few of them; (ii) the beta seems to be very sparse. Therefore, the user needs to generate diagnostic plots to assess the performance of the initial analysis, and only if needed, use an iterative approach to define better control genes. For this tutorial, we use HK genes as negative control genes.


## RUV4
RUV4 can be run with different values of k in an attempt to find the optimal value. Note that although there is a function to estimate k, called getK(), this may not give the optimal value for k and is often recomended to be used when there is no other choice but to automate finding k (e.g. in simulations). 

We begin with the list of housekeeping (HK) genes as our negative controls.
```{r}
HKgenes <- read.table("HouseKeeping_genes_IDs.txt", header=T, sep="\t")
hk <- HKgenes$GeneID
ctrl <- colnames(YA) %in% hk
head(ctrl)
```

We also define our biological factor of interest (i.e. control versus TGFb treatment)
```{r}
## Take treatment as the biological factor of interest in data set A
groups_A <- factor(info_datasetA$treatment) 
gA <- cbind(as.numeric(groups_A))   ## 1 control, 2: TGFb

## Take treatment as the biological factor of interest in data set B
groups_B <- factor(info_datasetB$treatment) 
gB <- cbind(as.numeric(groups_B))   ## 1 control, 2: TGFb

```

### Unadjusted data
Before running RUV4 normalisation with different k values, we begin by performing an analysis with no adjustment, i.e. k = 0. We then assess the variability between the two data sets A and B in the presence of unwanted variation.
```{r DE-unadjusted-A-B-using-HK}
# RUV4 with k = 0 for no adjustment
#----- In data set A 
fit_unadj_hk_datasetA <- RUV4(YA, X = gA,
                          ctl = ctrl,
                          k = 0)
fit_unadj_hk_datasetA.summary <- ruv_summary(YA,
                                        fit_unadj_hk_datasetA,
                                        info_datasetA)

kable(head(fit_unadj_hk_datasetA.summary$C, 2))

##----- In data set B 
fit_unadj_hk_datasetB <- RUV4(YB, X = gB,
                          ctl = ctrl,
                          k = 0)
fit_unadj_hk_datasetB.summary <- ruv_summary(YB,
                                         fit_unadj_hk_datasetB,
                                         info_datasetB)

```

### RUV4-adjusted data with different values of k
``` {r ruv4-HK-differentKs}
## Select different values of k
ks <- c( 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 15, 18, 20, 22, 23, 24)

## Define empty lists and use them in the for loop to store the results
fit_ruv4_hk_datasetA_all_k <- list()
fit_ruv4_hk_datasetA_all_k.summary <- list()
fit_ruv4_hk_datasetB_all_k <- list()
fit_ruv4_hk_datasetB_all_k.summary <- list()

for (K in ks){
  fit_ruv4_hk_datasetA_all_k[[K]] <- RUV4(YA, X = gA, 
                            ctl = ctrl, 
                            k = K, Z = 1, eta = NULL, 
                            fullW0 = NULL, inputcheck = TRUE)
  
  fit_ruv4_hk_datasetA_all_k.summary[[K]] <- ruv_summary(YA,
                                           fit_ruv4_hk_datasetA_all_k[[K]],
                                           info_datasetA)
  
  fit_ruv4_hk_datasetB_all_k[[K]] <- RUV4(YB, X = gB, 
                            ctl = ctrl, 
                            k = K,Z = 1, eta = NULL, 
                            fullW0 = NULL, inputcheck = TRUE)
  
  fit_ruv4_hk_datasetB_all_k.summary[[K]] <- ruv_summary(YB,
                                           fit_ruv4_hk_datasetB_all_k[[K]],
                                           info_datasetB)
}  
```

## Canonical correlation plot
In order to select an optimal value for k, we should plot, for each value of k, the square of the first canonical correlation between X and the first K left singular values of Y and Yc.

```{r}
ruv_cancorplot(YA, info_datasetA$treatment, ctrl) + ggtitle("dataset A")
ruv_cancorplot(YB, info_datasetB$treatment, ctrl) + ggtitle("dataset B")
```

To assess the consistency in the two data sets A and B, we computed the correlation between the betahats (logFC) of all genes in each data set.

```{r}
##----- Unadjusted data sets
plot(fit_unadj_hk_datasetA$betahat, 
     fit_unadj_hk_datasetB$betahat,
     xlab = "Betahat dataset A",
     ylab = "Betahat dataset B",
     main = "Unadjusted",
     xlim = c(-3,3), cex = 0.3, ylim = c(-4,4))
corVal <- cor.test(fit_unadj_hk_datasetA$betahat, 
                   fit_unadj_hk_datasetB$betahat)$estimate
text(-3,3, pos = 4, paste("Correlation: ", round(corVal, 3), sep = ""))

##------ RUV4-adjusted
for(K in ks) {
  plot(fit_ruv4_hk_datasetA_all_k[[K]]$betahat,
       fit_ruv4_hk_datasetB_all_k[[K]]$betahat,
       xlab = "Betahat dataset A",
       ylab = "Betahat dataset B",
       main = paste("RUV4_K_", K, sep=""),
       xlim = c(-3,3), cex = 0.3, ylim = c(-4,4))
  corVal <- cor.test(fit_ruv4_hk_datasetA_all_k[[K]]$betahat,
                     fit_ruv4_hk_datasetB_all_k[[K]]$betahat)$estimate
  text(-3,3, pos=4, paste("Correlation: ", round(corVal,3), sep=""))
}
```

Now, we choose k = 5 for dataset A and k = 3 for dataset B.
```{r}
plot(fit_ruv4_hk_datasetA_all_k[[5]]$betahat,
     fit_ruv4_hk_datasetB_all_k[[3]]$betahat,
     xlab = "Betahat dataset A",
     ylab = "Betahat dataset B",
     main = "RUV4_KA5_KB3",
     xlim = c(-3,3), cex = 0.3, ylim = c(-4,4))
  #abline(fit_ruv4_emp_datasetB$betahat,fit_ruv4_emp_datasetA$betahat)
  corVal <- cor.test(fit_ruv4_hk_datasetA_all_k[[K]]$betahat,
                     fit_ruv4_hk_datasetB_all_k[[K]]$betahat)$estimate
  text(-3,3, pos=4, paste("Correlation: ", round(corVal, 3),sep=""))
```

## P-value distributions and volcano plots for unadjusted and RUV4-adjusted data sets
Here we compare the results obtained from unadjusted and RUV4-adjusted data using p-value distributions. A good p-value distribution has a peak for low values and show uniform distribution for the larger values.

```{r}
KA = 5
KB = 3
##----- In data set A
ruv_hist(fit_unadj_hk_datasetA.summary) + ggtitle("Unadj_A")
ruv_hist(fit_ruv4_hk_datasetA_all_k.summary[[KA]]) + ggtitle("RUV4_HK_A")

## ---- In data set B
ruv_hist(fit_unadj_hk_datasetB.summary) + ggtitle("Unadj_B")
ruv_hist(fit_ruv4_hk_datasetB_all_k.summary[[KB]]) + ggtitle("RUV4_HK_B")
```
Compare the volcano plots between unadjusted and RUV4-adjusted data. A volcano plot visulaises the relationship between p-value and logFC for all genes.
```{r}
KA = 5
KB = 3
##----- In data set A
genecoloring <- list(aes(color = fit.ctl),
                     scale_color_manual(name = "Gene Category",
                                        values = alpha(c("Red", "black"),
                                                     c( 0.1, 0.25))))
ruv_volcano(fit_unadj_hk_datasetA.summary) + genecoloring + ggtitle("Unadj_A")
ruv_volcano(fit_ruv4_hk_datasetA_all_k.summary[[KA]]) + genecoloring + ggtitle("RUV4_HK_A")

##----- In data set A

ruv_volcano(fit_unadj_hk_datasetB.summary) + genecoloring + ggtitle("Unadj_B")
ruv_volcano(fit_ruv4_hk_datasetB_all_k.summary[[KB]]) + genecoloring + ggtitle("RUV4_HK_B")

```


## Overlapping DEGs between data sets A and B 
Finally, for each of the unadjusted and RUV4-adjusted data sets, we look at the number of overlapping differentially expressed genes (DEGs) between the two data sets A and B.\

First, we define DEGs as genes with adjusted p-value < 0.05 and |logFC| > 1 in data sets A and B which are unadjusted, RUV4-adjusted with K = 23, and RUV-4 adjusted with different values of K. Then we look at the overlapping genes in each data set.

### DEGs in the unadjusted data sets
Venn diagram comparing DEGs obtained from the unadjusted data sets A and B.
```{r Venn-DEGs-unadj}
##------ Define DEGs in the unadjusted data set A
DEGsUnadj_datasetA <- row.names(fit_unadj_hk_datasetA.summary$C)[fit_unadj_hk_datasetA.summary$C$F.p.BH < 0.05 & abs(fit_unadj_hk_datasetA.summary$C$b_X1) > 1]

##------ Define DEGs in the unadjusted data set B
DEGsUnadj_datasetB <- row.names(fit_unadj_hk_datasetB.summary$C)[fit_unadj_hk_datasetB.summary$C$F.p.BH < 0.05 & abs(fit_unadj_hk_datasetB.summary$C$b_X1) > 1]

##------ Explore how many genes are overlapping between data sets A and B
allDEGs_unadj <- c(DEGsUnadj_datasetA, 
                    DEGsUnadj_datasetB)

## remove duplicated gene symbols:
allDEGs_unadj <- allDEGs_unadj[!duplicated(allDEGs_unadj)]  

## Draw a Venn diagram comparing DEGs for RUVinv
Counts_unadj <- matrix(0, nrow = length(allDEGs_unadj), ncol = 2)
row.names(Counts_unadj)<- allDEGs_unadj
colnames(Counts_unadj)<- c("Unadj_A","Unadj_B")

for( i in 1:length(allDEGs_unadj)) {
  Counts_unadj[i,1]<- allDEGs_unadj[i] %in% DEGsUnadj_datasetA
  Counts_unadj[i,2]<- allDEGs_unadj[i] %in% DEGsUnadj_datasetB
}

col <- c("blue", "violet")
vennDiagram(vennCounts(Counts_unadj), 
            circle.col = col, 
            cex = c(1.6, 1.2, 1), lwd=2)

```


### DEGs in the RUV4-adjusted data sets for different values of k
Venn diagram comparing the RUV4-adjusted data sets A and B using different ks.
```{r Venn-DEGs-RUV4}

##---- Define DEGs for different values of k in data set A
DEGsRUV4_datasetA_all_k <- list()
for (K in ks){
  DEGsRUV4_datasetA_all_k[[K]] <- row.names(fit_ruv4_hk_datasetA_all_k.summary[[K]]$C)[fit_ruv4_hk_datasetA_all_k.summary[[K]]$C$F.p.BH < 0.05 & abs(fit_ruv4_hk_datasetA_all_k.summary[[K]]$C$b_X1) > 1]  
}

##---- Define DEGs for different values of k in data set B
DEGsRUV4_datasetB_all_k <- list()
for (K in ks){
  DEGsRUV4_datasetB_all_k[[K]] <- row.names(fit_ruv4_hk_datasetB_all_k.summary[[K]]$C)[fit_ruv4_hk_datasetB_all_k.summary[[K]]$C$F.p.BH < 0.05 & abs(fit_ruv4_hk_datasetB_all_k.summary[[K]]$C$b_X1) > 1]  
}

##---- Generate Venn diagrams, looking at the overlapping genes 
##     between data sets A and B for different values of k

allDEGs_RUV4_all_k <- list()

for (K in ks){
  allDEGs_RUV4_all_k[[K]] <- c(DEGsRUV4_datasetA_all_k[[K]],
                               DEGsRUV4_datasetB_all_k[[K]])
  ## remove duplicated gene symbols:
  allDEGs_RUV4_all_k[[K]] <- allDEGs_RUV4_all_k[[K]][!duplicated(allDEGs_RUV4_all_k[[K]])]  
  ## Draw a Venn diagram comparing DEGs for RUV4
  Counts_RUV4<- matrix(0, nrow = length(allDEGs_RUV4_all_k[[K]]), ncol=2)
  row.names(Counts_RUV4) <- allDEGs_RUV4_all_k[[K]]
  colnames(Counts_RUV4) <- c(paste("RUV4_A_K_", K, sep=""), paste("RUV4_B_K_",K,sep=""))

  for( i in 1:length(allDEGs_RUV4_all_k[[K]])) {
    Counts_RUV4[i,1]<- allDEGs_RUV4_all_k[[K]][i] %in% DEGsRUV4_datasetA_all_k[[K]]
    Counts_RUV4[i,2]<- allDEGs_RUV4_all_k[[K]][i] %in% DEGsRUV4_datasetB_all_k[[K]]
  }

  col<- c("blue", "violet")
  vennDiagram(vennCounts(Counts_RUV4), 
            circle.col = col, 
            cex = c(1.6, 1.2, 1), lwd = 2)
}


```

```{r}
KA <- 5
KB <- 3
allDEGs_RUV4 <- c(DEGsRUV4_datasetA_all_k[[KA]],
                             DEGsRUV4_datasetB_all_k[[KB]])
## remove duplicated gene symbols:
allDEGs_RUV4 <- allDEGs_RUV4[!duplicated(allDEGs_RUV4)]
## Draw a Venn diagram comparing DEGs for RUV4
Counts_RUV4<- matrix(0, nrow = length(allDEGs_RUV4), ncol=2)
row.names(Counts_RUV4) <- allDEGs_RUV4
colnames(Counts_RUV4) <- c("RUV4_KA5", "RUV4_KB3")

for( i in 1:length(allDEGs_RUV4)) {
  Counts_RUV4[i,1]<- allDEGs_RUV4[i] %in% DEGsRUV4_datasetA_all_k[[KA]]
  Counts_RUV4[i,2]<- allDEGs_RUV4[i] %in% DEGsRUV4_datasetB_all_k[[KB]]
}

col<- c("blue", "violet")
vennDiagram(vennCounts(Counts_RUV4),
          circle.col = col,
          cex = c(1.6, 1.2, 1), lwd = 2)

```

## Overlapping DEGs between unadjusted and RUV4-adjusted data sets
### Compare DEGs from the unadjusted and RUV4-adjusted data set A
```{r Venn-DEGs-datasetA}
KA <- 5
##----- Define DEGs for the RUV4-adjusted data with K = 5 in data set A
DEGsRUV4_datasetA <- row.names(fit_ruv4_hk_datasetA_all_k.summary[[KA]]$C)[fit_ruv4_hk_datasetA_all_k.summary[[KA]]$C$F.p.BH < 0.05 & abs(fit_ruv4_hk_datasetA_all_k.summary[[KA]]$C$b_X1) > 1]  

##----- Look at the overalapping genes between unadjusted and RUV-4 adjusted data set A
allDEGs_datasetA <- c(DEGsUnadj_datasetA,  
                      DEGsRUV4_datasetA_all_k[[KA]])

## remove duplicated gene symbols:
allDEGs_datasetA <- allDEGs_datasetA[!duplicated(allDEGs_datasetA)]  

## Draw a Venn diagram comparing DEGs for dataset A
Counts_datasetA <- matrix(0, nrow= length(allDEGs_datasetA), ncol=2)
row.names(Counts_datasetA)<- allDEGs_datasetA
colnames(Counts_datasetA)<- c("Unadj_A", "Ruv4_KA5")

for( i in 1:length(allDEGs_datasetA)) {
  Counts_datasetA[i,1]<- allDEGs_datasetA[i] %in% DEGsUnadj_datasetA
  Counts_datasetA[i,2]<- allDEGs_datasetA[i] %in% DEGsRUV4_datasetA_all_k[[KA]]
}

col<- c("blue","darkgreen")
vennDiagram(vennCounts(Counts_datasetA), 
            circle.col=col, 
            cex=c(1.6, 1.2, 1), lwd=2)

```

### Compare DEGs from the unadjusted and RUV4-adjusted data set B 
```{r Venn-DEGs-datasetB}
KB <- 3
##----- Define DEGs for the RUV4-adjusted data with K = 3 in data set B
DEGsRUV4_datasetB <- row.names(fit_ruv4_hk_datasetB_all_k.summary[[KB]]$C)[fit_ruv4_hk_datasetB_all_k.summary[[KB]]$C$F.p.BH < 0.05 & abs(fit_ruv4_hk_datasetB_all_k.summary[[KB]]$C$b_X1) > 1]  

##----- Look at the overalapping genes between unadjusted and RUV-4 adjusted data set B
allDEGs_datasetB <- c(DEGsUnadj_datasetB,
                      DEGsRUV4_datasetB_all_k[[KB]])

## remove duplicated gene symbols:
allDEGs_datasetB <- allDEGs_datasetB[!duplicated(allDEGs_datasetB)] 

## Draw a Venn diagram comparing DEGs for dataset B
Counts_datasetB <- matrix(0, nrow = length(allDEGs_datasetB), ncol = 2)
row.names(Counts_datasetB) <- allDEGs_datasetB
colnames(Counts_datasetB) <- c("Unadj_B", "Ruv4_KB3")

for( i in 1:length(allDEGs_datasetB)) {
  Counts_datasetB[i,1]<- allDEGs_datasetB[i] %in% DEGsUnadj_datasetB
  Counts_datasetB[i,2]<- allDEGs_datasetB[i] %in% DEGsRUV4_datasetB_all_k[[KB]]
}

col <- c("blue", "darkgreen")
vennDiagram(vennCounts(Counts_datasetB), 
            circle.col = col, 
            cex = c(1.6, 1.2, 1), lwd = 2)
```






