---
title: "How to remove unwanted variations from transcriptomics data using RUVinv and RUV4"
author:
- name: Sepideh Foroutan and Marie Trussart

output:
  html_notebook:
    number_sections: yes
    toc: yes
    toc_float: yes
    code_folding: "hide"
    fig_caption: yes
---


\
In this tutorial, we aim to normalise an integrated data generated by combining 10 microarray studies on control samples to and TGFb-treated samples extracted from this publication (Foroutan S. et al, "A Transcriptional Program for Detecting TGFÎ²-Induced EMT in Cancer", Molecular Cancer Research, 2017) that can be found here: http://mcr.aacrjournals.org/content/15/5/619.long.\ 

We first assess the presence of unwanted variations in the integrated data which combines differents studies and platforms. Second we show how to use two RUV methods: RUVinv and RUV4 to remove this unwanted variation and detect differential expressed genes comparing control samples to and TGFb-treated samples. Third we assess whether RUV methods are useful and compare the results obtained by each method.

# Data description

```{r load-libs}
library(ruv)
library(limma)          ## for DE analysis
#library(sva)            ## for performing ComBat and SVA
library(ruv)            ## for performing RUV
#library(NMF)            ## for drawing heatmaps
library(matrixStats)   
library(RColorBrewer)   ## for brewer.pal()
library(scales)         ## for alpha()
```


Add a sub-directory to save the output of the analysis.
```{r create-subdir}
mainPath <- getwd()
dir.create(file.path(mainPath, "out"), showWarnings = FALSE)
out_path <- "./out/"
```

We will be looking at two datasets sample A and sample B in this tutorial, each containing different platforms and studies. We will normalise them separately in order to compare the results obtained by the two normalisation methods.

Read in the integrated dataset:
```{r load-data}
samplesA<- read.table("expr_10data_sampleA.txt", header = T, sep = "\t")
samplesB<- read.table("expr_10data_sampleB.txt", header = T, sep = "\t")
```

Look at the data in each sample A and B:
```{r explore-data}

head(samplesA,3)
head(samplesB,3)
mA<-samplesA[,2:dim(samplesA)[2]]
mB<-samplesB[,2:dim(samplesB)[2]]
row.names(mA)<- samplesA[,1]
row.names(mB)<- samplesB[,1]
mA<- as.matrix(mA)
mB<- as.matrix(mB)
```

Look at the information related to each sample including the name of the studies, type of platforms, treatment, and tissue types:
```{r load-info-study}
info_samplesA<- read.table("info_10data_sampleA.txt", sep="\t", header=T)
info_samplesA
info_samplesB<- read.table("info_10data_sampleB.txt", sep="\t", header=T)
info_samplesB
```

# Assessment of unwanted variations in the data.
Here we perform some exploratory analysis on the integrated data to assess the amount of unwanted variation present in each dataset.

## RLE plot
We start by looking at the RLE plot on sample A by studies and by platform:
```{r rleplots-samplesA}
YA=t(mA)
### Plot RLE per study
ruv_rle(YA,info_samplesA$study,ylim = c(-4,4))
### Plot RLE per platform
ruv_rle(YA,info_samplesA$platform,ylim = c(-4,4))
```

Similarly, we look at the RLE plot on sample B by studies and by platform:
```{r rleplots-samplesB}
YB=t(mB)
### Plot RLE per study
ruv_rle(YB,info_samplesB$study,ylim = c(-4,4))
### Plot RLE per platform
ruv_rle(YB,info_samplesB$platform,ylim = c(-4,4))
```

## MDS plot
In transcriptomics applications, one of the most utilized exploratory plots is the multi-dimensional scaling (MDS) plot or a principal component analysis (PCA) plot. To assess the presence of unwanted variation in each sample, we use MDS plots to show similarities between samples measured in an unsupervised way. In transcriptomics, distances between samples are calculated based on the expression of the top varying genes. Ideally, samples should cluster well within the same treatment but we can identify in our MDS plot the unwanted variation as samples rather cluster by studies in both sample A and B.\
In the current example, it's important to note that in a few cases the platforms and studies are confounded, and therefore there is no way to identify how much of the unwanted variation come from the study and how much come from the platform. Such situations must be avoided when designing an experiment and for the purpose of this tutorial, we only consider "study" as the true source of variation.\

```{r mds-plot-sampleA}
library(ggplot2)
library(gridExtra)
gg_additions = list(aes(color=info_samplesA$study,shape=info_samplesA$treatment,size=5, alpha=.7),
labs(color="Study",shape="Treatment"),
scale_size_identity(guide="none"),
scale_alpha(guide="none"),
theme(legend.text=element_text(size=12),
legend.title=element_text(size=16)),
guides(color = guide_legend(override.aes = list(size = 4)),
shape = guide_legend(override.aes = list(size = 4)))
#scale_color_manual(values=c("darkorchid3", "darkorange2","dodgerblue3")
)
options(repr.plot.width=8, repr.plot.height=6)
ruv_svdplot(YA) + gg_additions #
```

```{r mds-plot-sampleB}
library(ggplot2)
library(gridExtra)
gg_additions = list(aes(color=info_samplesB$study,shape=info_samplesB$treatment,size=5, alpha=.7),
labs(color="Study",shape="Treatment"),
scale_size_identity(guide="none"),
scale_alpha(guide="none"),
theme(legend.text=element_text(size=12),
legend.title=element_text(size=16)),
guides(color = guide_legend(override.aes = list(size = 4)),
shape = guide_legend(override.aes = list(size = 4)))
#scale_color_manual(values=c("darkorchid3", "darkorange2","dodgerblue3")
)
options(repr.plot.width=8, repr.plot.height=6)
ruv_svdplot(YB) + gg_additions #
```


# Remove batch effects using RUVinv
There are several RUV methods for removing unwanted variations in order to obtain DEGs; these include RUV-1, RUV-2, RUV-4, RUV-inv and RUV-rinv. In general, RUV methods are dependent on negative control genes (genes which are not associated with the biological factor of interest) and replicate samples (if applicable). RUV-2 removes unwanted variations in two steps. RUV-4 came after RUV-2 and has four steps. For RUV-4 we need to estimate the number of unwanted variations (k), while for RUV-inv and RUV-rinv we don't need to estimate k as it is set to be maximum. In general, RUV-inv and RUV-rinv are better than RUV-4. RUV-inv is recommended when we have large number of control genes (~1000), while RUV-rinv is more appropriate with small number of control genes (~60). The negative control genes can be the housekeeping (HK) genes or emprical control genes.\

Here, we first focus on RUV-inv, which is performed by **RUVinv()** function. This function take expression matrix (Y), biological factor of interest (X), and a vector for indices of negative control genes (ctl).\
We load **ruv** package first and then we the list of housekeeping (HK) genes.

```{r}
# library(ruv)
HKgenes<- read.table("HouseKeeping_genes_IDs.txt", header=T, sep="\t")
hk<- HKgenes$GeneID
matA<- t(mA)
matB<- t(mB)
ctrl<- colnames(matA) %in% hk
```

We use these HK genes to perform RUV-inv and to obtain some statistics, from which we selected the genes that are mostly stable across the samples as the **empirical negative control genes** for sample A:

```{r ruv-inv-samplesA}
 ## take treatment as the biological factor of interest
groups_A<- factor(info_samplesA$treatment) 
gA<- cbind(as.numeric(groups_A))
## Apply RUV-inv using the housekeeping genes as negative controls to define new empirical controls genes
fit_samplesA=RUVinv(YA,gA,ctrl)
fit_samplesA.summary=ruv_summary(YA,fit_samplesA,info_samplesA)
head(fit_samplesA.summary$C)
### Selecting empirical negative controls genes
cGenes_samplesA <- row.names(fit_samplesA.summary$C)[fit_samplesA.summary$C$F.p.BH>0.05]  
empCtrl_ruvinv_samplesA<- colnames(matA) %in% cGenes_samplesA
Gene_Category=empCtrl_ruvinv_samplesA
Gene_Category[empCtrl_ruvinv_samplesA==T]="EmpCt"
Gene_Category[empCtrl_ruvinv_samplesA==F]="Non_EmpCt"
Emp_Ct_Genes_samplesA=data.frame(geneids=colnames(matA),empCt_RUV=empCtrl_ruvinv_samplesA,Categ=Gene_Category)
length(cGenes_samplesA)
```
 And similarly for sample B:
```{r ruv-inv-samplesB}
 ## take treatment as the biological factor of interest
groups_B<- factor(info_samplesB$treatment) 
gB<- cbind(as.numeric(groups_B))
## Apply RUV-inv using the housekeeping genesas negative controls to define new empirical controls genes
fit_samplesB=RUVinv(YB,gB,ctrl)
fit_samplesB.summary=ruv_summary(YB,fit_samplesB,info_samplesB)
head(fit_samplesB.summary$C)
### Selecting empirical negative controls genes
cGenes_samplesB <- row.names(fit_samplesB.summary$C)[fit_samplesB.summary$C$F.p.BH>0.05]  ## 7176
empCtrl_ruvinv_samplesB<- colnames(matB) %in% cGenes_samplesB
Gene_Category=empCtrl_ruvinv_samplesB
Gene_Category[empCtrl_ruvinv_samplesB==T]="EmpCt"
Gene_Category[empCtrl_ruvinv_samplesB==F]="Non_EmpCt"
Emp_Ct_Genes_samplesB=data.frame(geneids=colnames(matB),empCt_RUV=empCtrl_ruvinv_samplesB,Categ=Gene_Category)
length(cGenes_samplesB)
```

Now we use those empirical control genes to apply RUV-inv.
```{r emp-ruvinv}
fit_ruvin_emp_samplesA=RUVinv(Y=YA, X=gA, ctl=Emp_Ct_Genes_samplesA$empCt_RUV, Z = 1, fullW0 = NULL, lambda = NULL, iterN = 100000)
fit_ruvin_emp_samplesA.summary=ruv_summary(YA,fit_ruvin_emp_samplesA,info_samplesA,Emp_Ct_Genes_samplesA)
fit_ruvin_emp_samplesB=RUVinv(Y=YB, X=gB, ctl=Emp_Ct_Genes_samplesB$empCt_RUV, Z = 1, fullW0 = NULL, lambda = NULL, iterN = 100000)
fit_ruvin_emp_samplesB.summary=ruv_summary(YB,fit_ruvin_emp_samplesB,info_samplesB,Emp_Ct_Genes_samplesB)
```

Here, we look at the statistics obtained for sample A
```{r emp-ruvinv-sampleA-results}
## Look at the distribution of adjusted p-values
ruv_hist(fit_ruvin_emp_samplesA.summary)

## Look at the distribution of Volcano plots
genecoloring = list(aes(color=Categ),scale_color_manual(name="Gene Category",values=alpha(c("Black","Red"),c( 0.2, 0.2))))
ruv_volcano(fit_ruvin_emp_samplesA.summary)+genecoloring

## Look at ECDF of p-values
ruv_ecdf(fit_ruvin_emp_samplesA.summary) + genecoloring
```


Now we look at the statistics obtained for sample B
```{r emp-ruvinv-sampleB-results}
## Look at the distribution of adjusted p-values
ruv_hist(fit_ruvin_emp_samplesB.summary)

## Look at the distribution of Volcano plots
genecoloring = list(aes(color=Categ),scale_color_manual(name="Gene Category",values=alpha(c("Black","Red"),c( 0.2, 0.2))))
ruv_volcano(fit_ruvin_emp_samplesB.summary)+genecoloring

## Look at ECDF of p-values
ruv_ecdf(fit_ruvin_emp_samplesB.summary) + genecoloring
```


# Comparison of RUVinv method with RUV4 method
We want to assess the effect of our RUVinv method and compare it to the results obtained by RUV4.
Did we help ?\

## P-values distribution
To answer this question, we will run a RUV with k=0 for no adjustment and compare the results to RUVinv and RUV4, namely the pvalues distribution for the unadjusted method for both sample A and sample B, the pvalues distribution for the RUVinv method for both sample A and sample B and finally the pvalues distribution for the RUV4 method for both sample A and sample B.

``` {r comparison-methods}
### Did we help?
# RUV4 with k = 0 for no adjustment
# Equivalent to a Limma Analysis
fit_sampleA.unadj = RUV4(YA, X=gA, ctl=Emp_Ct_Genes_samplesA$empCt_RUV, 0)
fit_sampleA.summary.unadj = ruv_summary(YA,fit_sampleA.unadj,info_samplesA,Emp_Ct_Genes_samplesA)
fit_sampleB.unadj = RUV4(YB, X=gB, ctl=Emp_Ct_Genes_samplesB$empCt_RUV, 0)
fit_sampleB.summary.unadj = ruv_summary(YB,fit_sampleB.unadj,info_samplesB,Emp_Ct_Genes_samplesB)

## Applying RUV4
#Sample A
# Estimate best k
estimateK<- getK(YA, X=gA, ctl=Emp_Ct_Genes_samplesA$empCt_RUV, Z = 1, eta = NULL, fullW0 = NULL, cutoff = NULL,
                     method="select", l=1, inputcheck = TRUE)  
kA<- estimateK$k
fit_ruv4_emp_sampleA = RUV4(YA, X=gA, ctl=Emp_Ct_Genes_samplesA$empCt_RUV, k=kA,Z = 1, eta = NULL, fullW0 = NULL,
               inputcheck = TRUE)
fit_ruv4_emp_sampleA.summary = ruv_summary(YA,fit_ruv4_emp_sampleA,info_samplesA,Emp_Ct_Genes_samplesA)
#Sample B
# Estimate best k
estimateK<- getK(YB, X=gB, ctl=Emp_Ct_Genes_samplesB$empCt_RUV, Z = 1, eta = NULL, fullW0 = NULL, cutoff = NULL,
                     method="select", l=1, inputcheck = TRUE)  
kB<- estimateK$k
fit_ruv4_emp_sampleB = RUV4(YB, X=gB, ctl=Emp_Ct_Genes_samplesB$empCt_RUV, k=kB,Z = 1, eta = NULL, fullW0 = NULL,
               inputcheck = TRUE)
fit_ruv4_emp_sampleB.summary = ruv_summary(YB,fit_ruv4_emp_sampleB,info_samplesB,Emp_Ct_Genes_samplesB)


# Make a list of plots to compare side-by-side
#plots = list(
ruv_hist(fit_sampleA.summary.unadj)
ruv_hist(fit_sampleB.summary.unadj)
ruv_hist(fit_ruvin_emp_samplesA.summary)
ruv_hist(fit_ruvin_emp_samplesB.summary)
ruv_hist(fit_ruv4_emp_sampleA.summary)
ruv_hist(fit_ruv4_emp_sampleB.summary)
#)
```

## Betahat correlation
Another way is to compare how similar are the results obtained by the two datasets sample A and sample B after applying RUVinv and RUV4. To do so we look at the correlation between betahat from sample A and betahat from sample B for the unadjusted method, RUVinv and RUV4

```{r betahat-samples-A-B}
# unadjusted
plot(fit_sampleA.unadj$betahat,fit_sampleB.unadj$betahat,xlab="Betahat Samples A",ylab="Betahat Samples B",main="Unadjusted",xlim=c(-3,3),cex=0.3,ylim=c(-4,4))
#abline(fit_sampleB.unadj$betahat,fit_sampleA.unadj$betahat)
c1=cor.test(fit_sampleB.unadj$betahat,fit_sampleA.unadj$betahat)
text(-3,3, pos=4, paste("Correlation: ",round(c1$estimate,2),sep=""))
# RUVinv
plot(fit_ruvin_emp_samplesA$betahat,fit_ruvin_emp_samplesB$betahat,xlab="Betahat Samples A",ylab="Betahat Samples B",main="RUVinv",xlim=c(-3,3),cex=0.3,ylim=c(-4,4))
#abline(fit_ruvin_emp_samplesB$betahat,fit_ruvin_emp_samplesA$betahat)
c1=cor.test(fit_ruvin_emp_samplesA$betahat,fit_ruvin_emp_samplesB$betahat)
text(-3,3, pos=4, paste("Correlation: ",round(c1$estimate,2),sep=""))
# RUV4
plot(fit_ruv4_emp_sampleA$betahat,fit_ruv4_emp_sampleB$betahat,xlab="Betahat Samples A",ylab="Betahat Samples B",main="RUV4",xlim=c(-3,3),cex=0.3,ylim=c(-4,4))
#abline(fit_ruv4_emp_sampleB$betahat,fit_ruv4_emp_sampleA$betahat)
c1=cor.test(fit_ruv4_emp_sampleA$betahat,fit_ruv4_emp_sampleB$betahat)
text(-3,3, pos=4, paste("Correlation: ",round(c1$estimate,2),sep=""))
```

## Overlap between differential expressed genes obtained with RUVinv and RUV4
Finally we can also look at the differentially expressed genes obtained by each method RUVinv and RUV4 accross the two samples A and B and the overlap between methods using a Venn diagram
to check for consistency between methods on the same sample.

```{r compare_DEG_venndiag-samples}
## Obtain DEGs defined as having p-val<0.05 and abs(logFC)=abs(betahats)=1 for sample A:
DEGsRUVinv_sampleA<- row.names(fit_ruvin_emp_samplesA.summary$C)[fit_ruvin_emp_samplesA.summary$C$F.p.BH<0.05&abs(fit_ruvin_emp_samplesA.summary$C$b_X1)>1]  
DEGsRUV4_sampleA <- row.names(fit_ruv4_emp_sampleA.summary$C)[fit_ruv4_emp_sampleA.summary$C$F.p.BH<0.05&abs(fit_ruv4_emp_sampleA.summary$C$b_X1)>1]  
allDEGs_sampleA<- c(DEGsRUVinv_sampleA,DEGsRUV4_sampleA)
## remove duplicated gene symbols:
allDEGs_sampleA<- allDEGs_sampleA[!duplicated(allDEGs_sampleA)]  
## Draw a Venn diagram comparing DEGs for sample A
Counts_sampleA <- matrix(0, nrow= length(allDEGs_sampleA), ncol=2)
row.names(Counts_sampleA)<- allDEGs_sampleA
colnames(Counts_sampleA)<- c("RUVinv_A","Ruv4_A")
for( i in 1:length(allDEGs_sampleA)) {
  Counts_sampleA[i,1]<- allDEGs_sampleA[i] %in% DEGsRUVinv_sampleA
  Counts_sampleA[i,2]<- allDEGs_sampleA[i] %in% DEGsRUV4_sampleA
}
col<- c("blue", "violet")
vennDiagram(vennCounts(Counts_sampleA), circle.col=col, cex=c(1.6, 1.2, 1), lwd=2)

## Obtain DEGs defined as having p-val<0.05 and abs(logFC)=abs(betahats)=1 for sample B:
DEGsRUVinv_sampleB<- row.names(fit_ruvin_emp_samplesB.summary$C)[fit_ruvin_emp_samplesB.summary$C$F.p.BH<0.05&abs(fit_ruvin_emp_samplesB.summary$C$b_X1)>1]  
DEGsRUV4_sampleB <- row.names(fit_ruv4_emp_sampleB.summary$C)[fit_ruv4_emp_sampleB.summary$C$F.p.BH<0.05&abs(fit_ruv4_emp_sampleB.summary$C$b_X1)>1]  
allDEGs_sampleB<- c(DEGsRUVinv_sampleB,DEGsRUV4_sampleB)
## remove duplicated gene symbols:
allDEGs_sampleB<- allDEGs_sampleB[!duplicated(allDEGs_sampleB)]  
## Draw a Venn diagram comparing DEGs for sample B
Counts_sampleB <- matrix(0, nrow= length(allDEGs_sampleB), ncol=2)
row.names(Counts_sampleB)<- allDEGs_sampleB
colnames(Counts_sampleB)<- c("RUVinv_B","Ruv4_B")
for( i in 1:length(allDEGs_sampleB)) {
  Counts_sampleB[i,1]<- allDEGs_sampleB[i] %in% DEGsRUVinv_sampleB
  Counts_sampleB[i,2]<- allDEGs_sampleB[i] %in% DEGsRUV4_sampleB
}
col<- c("blue", "violet")
vennDiagram(vennCounts(Counts_sampleB), circle.col=col, cex=c(1.6, 1.2, 1), lwd=2)
```

## Overlap between differential expressed genes obtained across samples
We can also look at the differentially expressed genes obtained by each sample A and B for the same method RUVinv and RUV4 and the overlap between samples using a Venn diagram.

```{r compare_DEG_venndiag-methods}
## Obtain DEGs defined as having p-val<0.05 and abs(logFC)=abs(betahats)=1 for RUVinv
allDEGs_RUVinv<- c(DEGsRUVinv_sampleA,DEGsRUVinv_sampleB)
## remove duplicated gene symbols:
allDEGs_RUVinv<- allDEGs_RUVinv[!duplicated(allDEGs_RUVinv)]  
## Draw a Venn diagram comparing DEGs for RUVinv
Counts_RUVinv <- matrix(0, nrow= length(allDEGs_RUVinv), ncol=2)
row.names(Counts_RUVinv)<- allDEGs_RUVinv
colnames(Counts_RUVinv)<- c("RUVinv_A","RUVinv_B")
for( i in 1:length(allDEGs_RUVinv)) {
  Counts_RUVinv[i,1]<- allDEGs_RUVinv[i] %in% DEGsRUVinv_sampleA
  Counts_RUVinv[i,2]<- allDEGs_RUVinv[i] %in% DEGsRUVinv_sampleB
}
col<- c("blue", "violet")
vennDiagram(vennCounts(Counts_RUVinv), circle.col=col, cex=c(1.6, 1.2, 1), lwd=2)

## Obtain DEGs defined as having p-val<0.05 and abs(logFC)=abs(betahats)=1 for RUV4
allDEGs_RUV4<- c(DEGsRUV4_sampleA,DEGsRUV4_sampleB)
## remove duplicated gene symbols:
allDEGs_RUV4<- allDEGs_RUV4[!duplicated(allDEGs_RUV4)]  
## Draw a Venn diagram comparing DEGs for RUV4
Counts_RUV4 <- matrix(0, nrow= length(allDEGs_RUV4), ncol=2)
row.names(Counts_RUV4)<- allDEGs_RUV4
colnames(Counts_RUV4)<- c("RUV4_A","RUV4_B")
for( i in 1:length(allDEGs_RUV4)) {
  Counts_RUV4[i,1]<- allDEGs_RUV4[i] %in% DEGsRUV4_sampleA
  Counts_RUV4[i,2]<- allDEGs_RUV4[i] %in% DEGsRUV4_sampleB
}
col<- c("blue", "violet")
vennDiagram(vennCounts(Counts_RUV4), circle.col=col, cex=c(1.6, 1.2, 1), lwd=2)
```





