---
title: "RUV-III Challenges"
subtitle: "useR! 2018 - Brisbane"
output:
  github_document:
    toc: yes
    toc_depth: 3
  html_document:
    df_print: paged
    toc: yes
    toc_depth: 2
    toc_float: yes
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE)
```

The two challenges below are based on the RUV method for normalisation, RUV-III. There are two datasets provided containing gene expression data: `Challenge1Data.csv` and `Challenge2Data.csv`. The dataset `Challenge.information.csv` contains the clinical information of the samples (biological conditions of interest and gender of the samples).

There are 5 different biological conditions of interest (in `Challenge.information.csv` column `BiologicalConditions`) which have been run in 5 different batches. 

## Tasks

The two datasets are the same but one has already been normalised and the other one hasn't. 
Use `Challenge1Data.csv` or `Challenge2Data.csv` and assess if:

* The gene expression needs normalisation
* If the data needs normalisation, explain why and apply a normalisation.
* If you think the data has already been normalised, explain how you deduce it and comment on the quality of the normalisation.
* The same set of negative control genes is provided in both challenges. It is up to you to decide whether you need it!

**Hints**

* Remember that there is not one way to tell if the data needs normalisation
* Remember to check the known biology! Consider analysing the expression of the Y chromosome  genes `ZFY` and `EIF1AY` to inform your decision. These genes on the Y chromosome should be expressed in male samples but not in female samples.

## Challenge 1 data

```{r}
ch1_data <- read.csv("Challenge1Data.csv", row.names = 1, as.is = TRUE, stringsAsFactors = FALSE)

Con100 <- readRDS('controlGene100.rds')

Info <- read.csv('Challenge.information.csv', row.names = 1, as.is = TRUE, stringsAsFactors = FALSE)

ch1_data <- as.matrix(ch1_data)
kable(ch1_data[1:4,1:4])

kable(head(Info))

```

## Challenge 2 data

```{r}
ch2_data <- read.csv("Challenge2Data.csv", row.names = 1, as.is = TRUE, stringsAsFactors = FALSE)
Con100 <- readRDS('controlGene100.rds')
Info <- read.csv('Challenge.information.csv', row.names = 1, as.is = TRUE, stringsAsFactors = FALSE)

ch2_data <- as.matrix(ch2_data)
kable(ch2_data[1:4,1:4])

kable(head(Info))

```
